<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <!-- Gazebo 플러그인 설정 -->
    <gazebo reference="arm_base_link">
        <material>Gazebo/Orange</material>
    </gazebo>

    <gazebo reference="forearm_link">
        <material>Gazebo/Yellow</material>
    </gazebo>
    
    <gazebo reference="hand_link">
        <material>Gazebo/Orange</material>
    </gazebo>

    <ros2_control name="arm_robot_hardware" type="system">
        <hardware>
            <plugin>gazebo_ros2_control/GazeboSystem</plugin>
            <param name="parameters">$(find rescue_robot)/config/arm_control.yaml</param>
        </hardware>

        <joint name="arm_base_forearm_joint">
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>

        <joint name="forearm_hand_joint">
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>
    </ros2_control>

    <!-- ROS2 Control Plugin (gazebo_ros2_control) 사용 -->
    <gazebo>
        <plugin name="gazebo_ros2_control"
                filename="libgazebo_ros2_control.so">
            
            <parameters>$(find rescue_robot)/config/arm_control.yaml</parameters>
                <!-- Controller Manager 정의 -->
                
            <ros2_control>
                <controller_manager>controller_manager</controller_manager>
            </ros2_control>
        </plugin>
    </gazebo>

    <!-- 조인트 상태 업데이트 (joint_state_publisher) -->
    <gazebo>
        <plugin name="joint_state_publisher_controller"
                filename="libgazebo_ros_joint_state_publisher.so">
            <update_rate>10</update_rate>
            <joint_name>arm_base_forearm_joint</joint_name>
            <joint_name>forearm_hand_joint</joint_name>
        </plugin>
    </gazebo>

    <!-- Gazebo에서 제어할 조인트 설정 -->
    <gazebo>
        <plugin name="joint_trajectory_controller"
                filename="libgazebo_ros_joint_pose_trajectory.so">
            <update_rate>10</update_rate>  <!-- 업데이트 속도 -->

            <frame_id>base_footprint</frame_id> <!-- 기준 프레임 -->

            <!-- 제어할 조인트들 -->
            <controlled_joints>
                <joint>arm_base_forearm_joint</joint>
                <joint>forearm_hand_joint</joint>
            </controlled_joints>
        </plugin>
    </gazebo>

</robot>
